### PR TITLE
[cudagraph] fallback to eager if re-record too many times

### PR DESCRIPTION
**Pull request by BoyuanFeng**
Summary:
CUDAGraph Trees previously relies on an assumption that static inputs (parameters and buffers) does not change tensor addresses across multiple function invocations. This assumption can be used to reduce the number of tensor copies to improve performance. We also use `check_static_inputs_are_stable()` to check whether this assumption holds at runtime.

While this assumption is True in most cases, we recently observe a few cases that this assumption is not valid:
- [Inline inbuilt nn modules](https://github.com/pytorch/pytorch/pull/126822): the same function (a nn module) is used in multiple places and different parameters and buffers are passed to this function with different tensor addresses
- Some user code changes tensor addresses of parameters/buffers. See [internal example]( https://www.internalfb.com/mlhub/pipelines/runs/mast/sw-935450288-OfflineTraining_08ba1cf0?job_attempt=1&version=0&env=PRODUCTION)
- Compiled Autograd may also pass parameters/buffers with different tensor addresses across runs.

Previous PR [#126822](https://github.com/pytorch/pytorch/pull/126822) (by @mlazos) allows detecting static tensor address changes during runtime and re-recording a cudagraph if that happened. However, if the same function is re-recorded too many times, it may introduce large overhead and hurt performance. This PR adds `torch._inductor.config.triton.cudagraph_max_recording` (=5) to fallback to eager if a function has been recorded more than `cudagraph_max_recording` times for a specific node in the CUDAGraph Trees.

A summary on how static tensor address changes are handled now:

- For each child node, check the assumption via `check_invariants`. If this holds, execute node with the assumption.
- If the assumption does not hold for all child nodes, re-record if the function_id has not been recorded too many times for the current_node.
- If the function_id has been re-recorded too many times, fallback to eager function and warning.

Test Plan: CI

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/cudagraph_trees.py [hunk 1]
```diff
@@ -754,10 +754,12 @@ def __init__(
...
         self.stack_traces = stack_traces
         self.stream = stream
-        # If we are inlining builtin nn modules we will re-record if static inputs change
-        # if not we should error because dynamo should have recompiled in this case
+
+        # Enable re-record a cudagraph when static tensor address changed.
+        # if not we should error when it changed.
         self.rerecord_if_static_inputs_change = (
             torch._dynamo.config.inline_inbuilt_nn_modules
+            or torch._inductor.config.cudagraph_support_input_mutation
```
Comment (mlazos) [inline torch/_inductor/cudagraph_trees.py:762]: Does this work without enabling inline_inbuilt_nn_modules? I don't think it will because AOT autograd needs to mark the correct argument indices as static.

Can you point me to the examples you guys are running into? I had a more narrow solution that involves marking only certain params as static in dynamo.

**Comment by mlazos** [comment]
@BoyuanFeng and I discussed this today. The outline of the problem is this:

There's a specific internal model that is now running cudagraphs which was not running it before due to Boyuan's change to allow input mutation when the input is cudagraph managed. This model specifically modifies parameter pointers on each iteration. 

My thoughts: 
I think in this case my ideal solution is to just disable cudagraphs. This model specifically is not fit for it since it modifies the locations of parameters on every iteration, so a copy of these will likely destroy any perf gain from using cudagraphs. However from our discussion it appears that this team does not want to have model-specific configurations for torch.compile. This PR circumvents this in essence by enabling re-recording of cudagraphs on static inputs and relying on a fallback to disable cudagraphs in order to reach the end-goal of disabling cudagraphs on this model. This seems kind of roundabout to me. 

Alternatively I feel like dynamo should actually be recompiling in this case and am surprised it isn't, although I suspect this is because the ID of the nn module is the same even though the underlying param pointers are changing (we don't guard on nn module internals due to performance concerns) This would also be bad for perf since we will recompile on every iteration, so this is a non-solution. 

The final thought I have is that I don't think blanket enabling cudagraphs in cases where it was not running it before is a good mode of operation. Perhaps blocking boyuan's mutation change behind a flag and disabling it for this cohort of models should be our approach (or is that change specifically to enable it for some models for this team?)

Potential solutions:
1. Block the cudagraph-owned input mutation feature by a flag for this cohort of models
2. Disable cudagraphs on this model
3. A fallback wrapper outside of torch.compile for this cohort of models (which tries different configurations)
4. This PR (allow re-recording for a small number of graphs, make this limit higher if we are inlining nn modules)
5. Could the model possibly be changed to not reassign parameter pointers?

**Comment by eellison** [comment]
@mlazos 

I agree with you that yes, probably this model that has changing parameter data ptrs is not super beneficial to cudagraph and they shouldn't cudagraph it. 

But, the larger context here is that we have a ton of models internally. It's going to make it super hard to iterate on features if every change causes models to break and then requires a per-model disabling. In this case, it was allowing mutated non-param inputs, but in the future it might be graph partitioning to avoid cpu inputs, etc. 
 
@boyuanfeng did in fact update one model to replace `self.attr = new_attr` with `self.attr.copy_(new_attr)`, and it still broke (potentially bc of out of place optimizer ?) Would you give some more details on that ? (Can msg both of us internally if it is giving too many details). 

It would be great if we could collect more details on the reason we are re-recording. Could we augment the warning msg with particular reasons we are re-recording ? In the future, maybe if there's only a single param that's changing we would just copy. Related, we could add a user-level apis around copying particular tensors. 

All of this also applies to inputs from prior cudagraph pools as well. 

Note: we should not be falling back to non-cudagraph for symint changes, or for recompiling control flow,  but i think that is is the case with this pr today.

**Comment by mlazos** [comment]
> @mlazos
> 
> I agree with you that yes, probably this model that has changing parameter data ptrs is not super beneficial to cudagraph and they shouldn't cudagraph it.
> 
> But, the larger context here is that we have a ton of models internally. It's going to make it super hard to iterate on features if every change causes models to break and then requires a per-model disabling. In this case, it was allowing mutated non-param inputs, but in the future it might be graph partitioning to avoid cpu inputs, etc.
> 
> @BoyuanFeng did in fact update one model to replace `self.attr = new_attr` with `self.attr.copy_(new_attr)`, and it still broke (potentially bc of out of place optimizer ?) Would you give some more details on that ? (Can msg both of us internally if it is giving too many details).
> 
> It would be great if we could collect more details on the reason we are re-recording. Could we augment the warning msg with particular reasons we are re-recording ? In the future, maybe if there's only a single param that's changing we would just copy. Related, we could add a user-level apis around copying particular tensors.
> 
> All of this also applies to inputs from prior cudagraph pools as well.
> 
> Note: we should not be falling back to non-cudagraph for symint changes, or for recompiling control flow, but i think that is is the case with this pr today.

What is worrisome to me is that we were skipping before and now we're running on more models that we weren't running on before, not necessarily that we're breaking existing models which were already running cudagraphs. Shouldn't the rollout be more systematic for a feature like this? I feel like this PR will paper over issues that we should be aware of (perhaps printing like you suggested could alleviate this). 

In the end if we don't have a better way I'm okay with this PR, I just felt that it's roundabout to getting to the end goal of disabling on this model, and it could possibly paper over issues that we should know about. 

My solution ranking is:
1) disable model in config
2) change code to not rewrite pointers
3) fallback wrapper outside of torch.compile
4) This PR

If 1-3 aren't feasible let's go ahead with this PR.

**Comment by eellison** [comment]
cc @dshi7 @xw285cornell @ckluk2 for thoughts re: production

File: torch/_inductor/config.py [hunk 2]
```diff
@@ -677,6 +677,11 @@ class triton:
     # Enable cudagraph support for mutated inputs from prior cudagraph pool
     cudagraph_support_input_mutation = False if is_fbcode() else True
 
+    # Maximal number of allowed cudagraph re-record for a function and
+    # a cudagraph node due to static input tensor address changes.
+    # i.e., allow num_recording <= cudagraph_max_rerecording_due_to_static_input_idx_mismatch
+    cudagraph_max_rerecording_due_to_static_input_idx_mismatch = 5
```
Comment (eellison) [inline torch/_inductor/config.py:683]: `cudagraph_static_input_rerecord_limit` ?

File: torch/_inductor/cudagraph_utils.py [hunk 3]
```diff
@@ -178,3 +179,17 @@ def get_placeholder_stack_trace(placeholder: torch.fx.Node) -> Optional[str]:
...
 
     return None
+
+
+class CheckInvariantStatus(Enum):
+    # Check invariant succeeded
+    SUCCESS = 1
+
+    # Previously managed data pointers are not stable
+    CudagraphManagedIdxMismatch = 2
```
Comment (eellison) [inline torch/_inductor/cudagraph_utils.py:189]: nice! someone has actually run into this previously and it was inscrutable to them. could we add this to the pr ?

File: torch/_inductor/cudagraph_trees.py [hunk 4]
```diff
@@ -835,9 +838,11 @@ def __init__(
 
         # precompute expanded dims to avoid computing in the hot path
         self.expanded_dims: List[List[int]] = [
-            get_expanded_dims(x)
-            if isinstance(x, torch.Tensor) and idx not in self.static_input_idxs
-            else []
+            (
```
Comment (eellison) [inline torch/_inductor/cudagraph_trees.py:841]: nit: unrelated syntax change

File: torch/_inductor/cudagraph_trees.py [hunk 5]
```diff
@@ -1972,6 +1994,23 @@ def _run(self, new_inputs: List[Tensor], function_id: FunctionID):
             if self.current_node is not None:
                 self.apply_checkpoint_execution_state_in_allocator()
 
+            # re-record due to static input tensor address changes
```
Comment (eellison) [inline torch/_inductor/cudagraph_trees.py:1997]: If we have 

node1 ---> node2

And we hit the limit on node2, then we'll be applying the checkpointing state on the allocator in each iteration. applying the checkpointing state in the allocator is slow. We checkpoint the allocator state because we have tensors in the cudagraph caching allocator pool which are unallocated, and we want to do more allocations to it. Here, we're about to do more allocations in the eager pool so we dont need to checkpoint it.

File: torch/_inductor/cudagraph_trees.py [hunk 6]
```diff
@@ -1569,7 +1576,7 @@ def _allocate_and_copy_recording_inputs(
 
         return recording_inputs
 
-    def check_invariants(self, inputs: List[Tensor]) -> bool:
+    def check_invariants(self, inputs: List[Tensor]) -> CheckInvariantStatus:
```
Comment (eellison) [inline torch/_inductor/cudagraph_trees.py:1579]: Could we return both the `CheckInvariantStatus` and also a function which would generate a more detailed message (to be used lazily) ?

Would be great to add additional info with stack traces of which inputs changed

File: torch/_inductor/cudagraph_trees.py [hunk 7]
```diff
@@ -1972,6 +1994,23 @@ def _run(self, new_inputs: List[Tensor], function_id: FunctionID):
...
+                curr_node_id = self._get_node_id()
+                self.num_rerecord[curr_node_id][function_id] += 1
+                if (
+                    self.num_rerecord[curr_node_id][function_id]
+                    > torch._inductor.config.triton.cudagraph_max_rerecording_due_to_static_input_idx_mismatch
+                ):
+                    _id = curr_node_id.id if curr_node_id else None
+                    log_cudagraph_skip_and_bump_counter(
+                        f"skipping cudagraph due to function {function_id.id} exceeding max "
+                        f"re-recording limit "
```
Comment (eellison) [inline torch/_inductor/cudagraph_trees.py:2008]: Nit: exceeding `cudagraph_static_input_rerecord_limit {value}`

File: torch/_inductor/cudagraph_trees.py [hunk 8]
```diff
@@ -1997,6 +2036,13 @@ def shutdown(self):
         self.current_node = None
 
     def record_function(self, new_inputs, function_id) -> List[Optional[Tensor]]:
+        curr_node_id = self._get_node_id()
+        if (
+            self.num_rerecord[curr_node_id][function_id]
+            > torch._inductor.config.triton.cudagraph_max_rerecording_due_to_static_input_idx_mismatch
+        ):
+            return self.ids_to_funcs[function_id].model(new_inputs)
```
Comment (eellison) [inline torch/_inductor/cudagraph_trees.py:2044]: Do we need both this and the other check above in run ? will we ever hit this path ?

**Review by eellison — COMMENTED** [review]
Looks great ! a couple comments.

There was an earlier issue of someone re-recording because of changing input location. Maybe as a follow up (or in this pr), it would be great to add a similar warning/limit, with input stack trace attribution.

See, post here: https://pytorch.org/blog/accelerating-generative-ai-4/#2-memory-pointer-management

File: torch/_inductor/config.py [hunk 9]
```diff
@@ -677,6 +677,12 @@ class triton:
     # Enable cudagraph support for mutated inputs from prior cudagraph pool
     cudagraph_support_input_mutation = False if is_fbcode() else True
 
+    # Maximal number of allowed cudagraph re-record for a function and
+    # a cudagraph node due to static input tensor address changes or
+    # cudagraph managed tensor data pointer changed.
+    # i.e., allow num_recording <= cudagraph_unexpected_rerecord_limit
+    cudagraph_unexpected_rerecord_limit = 5
```
[the remark this answers is not in the document]
Comment (eellison) [reply torch/_inductor/config.py:684]: The other thing we could do, is if we detect that a certain cudagraph pool input is repeatedly changing, copy it instead of specializing.

**Review by eellison — APPROVED** [review]
looks great! maybe we should start with a bit higher `cudagraph_unexpected_rerecord_limit`, but other than that, looks good !

File: torch/_inductor/config.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/_inductor/config.py:684]: I had naively expected that one would be willing to record a separate cuda graph per parameter set for an inlined nn module, where did I get this calculus wrong?

[the remark this answers is not in the document]
Comment (ezyang) [reply torch/_inductor/config.py]: Ah yes, out of place optimizer is a big problem. In fact, it is probably is so bad you might specifically try to rewrite your optimizer to be inplace, or double buffered or something lol
